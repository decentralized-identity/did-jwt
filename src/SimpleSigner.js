import { ec as EC } from 'elliptic'
import { sha256 } from './Digest'

const secp256k1 = new EC('secp256k1')

function leftpad (data, size = 64) {
  if (data.length === size) return data
  return '0'.repeat(size - data.length) + data
}
/**
*  The SimpleSigner returns a configured function for signing data. It also defines
*  an interface that you can also implement yourself and use in our other modules.
*
*  @example
*  const signer = SimpleSigner(process.env.PRIVATE_KEY)
*  signer(data, (err, signature) => {
*    ...
*  })
*
*  @param    {String}         hexPrivateKey    a hex encoded private key
*  @return   {Function}                     a configured signer function
*/

export default function (hexPrivateKey) {
  const privateKey = secp256k1.keyFromPrivate(hexPrivateKey)
  return async (data) => {
    const {r, s, recoveryParam} = privateKey.sign(sha256(data))
    return {
      r: leftpad(r.toString('hex')),
      s: leftpad(s.toString('hex')),
      recoveryParam
    }
  }
}
