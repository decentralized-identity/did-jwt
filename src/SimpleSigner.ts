import { ec as EC, ec } from 'elliptic'
import base64url from 'uport-base64url'
import { sha256 } from './Digest'
import { Signer, EcdsaSignature } from './JWT'

const secp256k1: EC = new EC('secp256k1')

function leftpad(data: string, size = 64): string {
  if (data.length === size) return data
  return '0'.repeat(size - data.length) + data
}

export function ecdsaSigToJose({ r, s, recoveryParam }: EcdsaSignature): string {
  const jose: Buffer = Buffer.alloc(65)
  Buffer.from(r, 'hex').copy(jose, 0)
  Buffer.from(s, 'hex').copy(jose, 32)
  jose[64] = recoveryParam
  return base64url.encode(jose)
}

/**
 *  The SimpleSigner returns a configured function for signing data. It also defines
 *  an interface that you can also implement yourself and use in our other modules.
 *
 *  @example
 *  const signer = SimpleSigner(process.env.PRIVATE_KEY)
 *  signer(data, (err, signature) => {
 *    ...
 *  })
 *
 *  @param    {String}         hexPrivateKey    a hex encoded private key
 *  @return   {Function}                     a configured signer function
 */

function SimpleSigner(hexPrivateKey: string): Signer {
  const privateKey: ec.KeyPair = secp256k1.keyFromPrivate(hexPrivateKey)
  return async data => {
    const { r, s, recoveryParam }: EC.Signature = privateKey.sign(sha256(data))
    const signatureObj: EcdsaSignature = {
      r: leftpad(r.toString('hex')),
      s: leftpad(s.toString('hex')),
      recoveryParam
    }
    return ecdsaSigToJose(signatureObj)
  }
}

export default SimpleSigner
