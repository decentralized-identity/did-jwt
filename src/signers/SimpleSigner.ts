import { fromJose } from '../util'
import { Signer } from '../JWT'
import { ES256KSigner } from './ES256KSigner'

/**
 * @deprecated Please use ES256KSigner
 *  The SimpleSigner returns a configured function for signing data.
 *
 *  @example
 *  const signer = SimpleSigner(process.env.PRIVATE_KEY)
 *  signer(data, (err, signature) => {
 *    ...
 *  })
 *
 *  @param    {String}         hexPrivateKey    a hex encoded private key
 *  @return   {Function}                     a configured signer function
 */
function SimpleSigner(hexPrivateKey: string): Signer {
  const signer = ES256KSigner(hexPrivateKey, true)
  return async (data) => {
    const signature = (await signer(data)) as string
    return fromJose(signature)
  }
}

export default SimpleSigner
