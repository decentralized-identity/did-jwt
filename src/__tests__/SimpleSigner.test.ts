import SimpleSigner from '../signers/SimpleSigner'

const privateKey = '278a5de700e29faae8e40e366ec5012b5ec63d36ec77e8a2417154cc1d25383f'
const signer = SimpleSigner(privateKey)
it('signs data', async () => {
  expect.assertions(1)
  const plaintext = 'thequickbrownfoxjumpedoverthelazyprogrammer'
  return await expect(signer(plaintext)).resolves.toMatchSnapshot()
})

const privateKey0x = '0x278a5de700e29faae8e40e366ec5012b5ec63d36ec77e8a2417154cc1d25383f'
it('signs data: privateKey with 0x prefix', async () => {
  expect.assertions(1)
  const signer2 = SimpleSigner(privateKey0x)
  const plaintext = 'thequickbrownfoxjumpedoverthelazyprogrammer'
  return await expect(signer2(plaintext)).resolves.toMatchSnapshot()
})
