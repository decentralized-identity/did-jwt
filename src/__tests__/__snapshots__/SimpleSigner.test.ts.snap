// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signs data 1`] = `
Object {
  "r": "8ecbdd2f0aabf8edb4ea191e828abaa5ba3b2c98c269f9442870af7e88413fd5",
  "recoveryParam": 0,
  "s": "c9e02ee61b64cf2fd623c8a1296125eaaa452b7d1a02c5079c1d1d75521ecf3b",
}
`;

exports[`signs data: privateKey with 0x prefix 1`] = `
Object {
  "r": "8ecbdd2f0aabf8edb4ea191e828abaa5ba3b2c98c269f9442870af7e88413fd5",
  "recoveryParam": 0,
  "s": "c9e02ee61b64cf2fd623c8a1296125eaaa452b7d1a02c5079c1d1d75521ecf3b",
}
`;
